### PR TITLE
Multiple API Keys per User

### PR DESCRIPTION
**Pull request by proteusvacuum**
Pass One from: https://trello.com/c/efXMl0ik/28-api-key-improvement-updates

##### SUMMARY

This adds the ability for a user to have multiple named API Keys. It integrates the IP whitelist directly into this model, and opens the door for future API Key work.

I have left the old `APIKeySettings` model and the tastypie `APIKey` migrations intact as a failsafe, and will remove those in a later PR. 

##### RISK ASSESSMENT / QA PLAN

I will submit a QA request after a first review on this PR.

##### PRODUCT DESCRIPTION

This adds a new "API Keys" page to the settings menu where a user can manage their API keys. 

Screenshot from 2020-06-01 12-07-57

### CODE REVIEW

File: corehq/apps/users/migrations/0016_hqapikey.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('key', models.CharField(blank=True, db_index=True, default='', max_length=128)),
+                ('name', models.CharField(blank=True, default='', max_length=255)),
+                ('created', models.DateTimeField(default=django.utils.timezone.now)),
+                ('ip_whitelist', django.contrib.postgres.fields.ArrayField(base_field=models.GenericIPAddressField(), default=list, size=None)),
+                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='api_keys', to=settings.AUTH_USER_MODEL)),
+            ],
+        ),
+        migrations.RunPython(populate_api_keys, migrations.RunPython.noop)
```
Comment (proteusvacuum) [inline corehq/apps/users/migrations/0016_hqapikey.py:50]: I wasn't sure what a down migration would look like for this... Re-populate the tastypie APIKey model?

Comment (calellowitz) [reply corehq/apps/users/migrations/0016_hqapikey.py:50]: This seems right

**Review by calellowitz — APPROVED** [review]
Not blocking, but this seems like a pretty big change in how ip whitelisting is managed. This changes the whitelist from something an admin sets to something the individual user sets. Is that what we want? A control the user can turn off does not seem like a useful control. At worst it feels like domain admins should be able to set these for other users in the domain, as part of access controls, whether that happens at a key or domain level, im not sure matters.

Not sure if @ctsims has more thoughts on that.

**Comment by proteusvacuum** [comment]
@calellowitz One thing to note is that you cannot edit an API Key whitelist once it has been created (other than through the admin), although I do agree that this changes the spirit of how this was managed previously. 
I'll wait for @ctsims to weigh in before pushing this to staging. Removing that field from the create form should be pretty easy.

**Comment by ctsims** [comment]
@calellowitz I think we want the IP whitelist in user control. all the IP whitelist allows you to do is **restrict permissions** from what the key already does out of the box, so I don't see a reason why it shouldn't be accessible to the user who creates the token. The only reason it's in the admin now is because it was the fastest execution mechanism.I think not letting you change it after creation is the control mechanism that prevents escalation. 

If there's an "Admin" level change in the future it should likely be linked with moving API keys from being PAT's to being "system" api keys, which can have different rules/roles for who can control them.

**Comment by orangejenny** [comment]
fyi, I added 143d9af7e24 to increment the migration now that https://github.com/dimagi/commcare-hq/pull/27763 is merged

**Comment by proteusvacuum** [comment]
QA passed. Would love another quick review on the latest changes :bowing_man:

**Review by orangejenny — APPROVED** [review]
Just looked at last 4 commits, looks good.

**Comment by czue** [comment]
@proteusvacuum @calellowitz do you know whether this PR invalidates the `ApiKeySettings` model that was added here: https://github.com/dimagi/commcare-hq/pull/27492/? I think we can delete it, but wanted to check if you knew something I didn't.

**Comment by proteusvacuum** [comment]
It does. We can delete that model now.

**Pull request merged**